NumberInput validation

achampagnedev opened the pull request:
### Project organization

- Related to https://github.com/liteflow-labs/marketplace-template/issues/72
- Related https://trello.com/c/eErTVlGD

MARKETPLACE TEMPLATE PR:
https://github.com/liteflow-labs/marketplace-template/pull/89/

### Description

Adding validation for number inputs throughout the library

### How to test

This is currently partially testable in storybook, some are gated behind the login flow and can be tested on this PR https://github.com/liteflow-labs/marketplace-template/pull/89/ : 

Components should be testable They are: 
- [ ] Offer -> Form -> Bid -> Default
- [ ] Offer -> Checkout -> Multiple
- [ ] Sales -> Auction -> Form -> Default
- [ ] Sales -> Direct -> Form -> Is Single / Is Multiple
- [ ] Token -> Form -> Create -> Is Single / Is Multiple

Templates (doesn't build right now therefore cannot be tested). Validate functions should be okay but I added the display of the errors in the UI and since I can't preview it, it's not tested right now. This could be tested in `marketplace-template` repo with PR https://github.com/liteflow-labs/marketplace-template/pull/89/ :
- [ ] Explore page -> Filters -> Min/Max price

To test for price inputs: 
- [ ] Value need to be positive
- [ ] Value can't exceed maximum decimal of currency
- [ ] Value cannot be empty unless optional

To test for quantity inputs:

- [ ] Value need to be positive
- [ ] Value need to be a minimum of 1
- [ ] Value can't exceed the maximum available quantity
- [ ] Value need to be integers

To test for royalty input:

- [ ] Value can be optional or 0
- [ ] Value need to be positive
- [ ] Value cannot exceed the maximum allowed royalty
- [ ] Is there a maximum decimal ? not for now.

### Checklist

- [x] Update related changelogs

achampagnedev commented on packages/components/src/Offer/Summary.tsx:
```diff
@@ -19,7 +19,12 @@ const Summary: FC<
   const totalPrice = useMemo(() => {
     if (!quantity) return BigNumber.from(0)
-    return price.mul(quantity)
+    try {
```
inspired from PR: https://github.com/liteflow-labs/liteflow-js/pull/68/files

achampagnedev commented on packages/templates/src/home/Explorer.tsx:
```diff
@@ -531,26 +555,57 @@ export const Template: VFC<
                     </NumberInput>
                   </InputGroup>
-                </Flex>
+                  <Stack spacing={1}>
```
Need to test UI for this.

achampagnedev commented on packages/templates/src/home/Explorer.tsx:
```diff
@@ -531,26 +555,57 @@ export const Template: VFC<
-                        {...register('maxPrice')}
+                        {...register('maxPrice', {
+                          validate: (value) => {
+                            if (!value) return
```
Need to test validation for both minPrice and maxPrice

achampagnedev commented on packages/templates/src/home/Explorer.tsx:
```diff
@@ -504,22 +505,45 @@ export const Template: VFC<
                         placeholder={t('explore.form.min-price.placeholder')}
-                        {...register('minPrice')}
+                        {...register('minPrice', {
+                          validate: (value) => {
```
Need to test validation for both minPrice and maxPrice

achampagnedev commented on packages/components/src/Sales/Direct/Form.tsx:
```diff
@@ -263,6 +271,12 @@ const SalesDirectForm: VFC<Props> = ({
                 {...register('quantity', {
                   required: t('sales.direct.form.validation.required'),
+                  validate: (value) =>
+                    (parseFloat(value) >= 1 &&
```
will replace this one for `if/else`

NicolasMahe commented on packages/components/src/Offer/Form/Checkout.tsx:
```diff
@@ -151,6 +149,21 @@ const OfferFormCheckout: FC<Props> = ({
+                        available: parseInt(offer.availableQuantity, 10),
+                      })
+                    }
+                  },
```
Would be a lot better to do the conditions to check for wrong value rather than checking for good value and have to cascade them together:
```suggestion
                  validate: (value) => {
                    if (!/^\d+$/.test(value)) {
                      return t('offer.form.checkout.validation.integer')
                    }
                    if (
                      parseInt(value, 10) < 0 ||
                      parseInt(value, 10) > parseInt(offer.availableQuantity, 10)
                    ) {
                      return t('offer.form.checkout.validation.in-range', {
                        available: parseInt(offer.availableQuantity, 10),
                      })
                    }
                  },
```

NicolasMahe commented on packages/components/src/Sales/Auction/Form.tsx:
```diff
@@ -165,7 +163,21 @@ const SalesAuctionForm: VFC<Props> = ({
+                      }
+                    } else {
+                      return t('sales.auction.form.validation.positive')
+                    }
```
Same comment about doing the condition in the other way:
```suggestion
                    if (parseFloat(value) <= 0)
                      return t('sales.auction.form.validation.positive')
                    if (splitValue[1].length > currency.decimals)
                      return t('sales.auction.form.validation.decimals', {
                        nbDecimals: currency.decimals,
                      })
```
Also `splitValue[1]` may be undefined if the number doesn't contain a `.`:
```
if (splitValue[1] && splitValue[1].length > currency.decimals)
```

NicolasMahe commented on packages/components/src/Sales/Direct/Form.tsx:
```diff
@@ -263,6 +271,12 @@ const SalesDirectForm: VFC<Props> = ({
+                      parseFloat(value) <= quantityAvailable?.toNumber()) ||
+                    t('sales.direct.form.validation.in-range', {
+                      quantityAvailable: quantityAvailable?.toNumber(),
+                    }),
```
Please avoid to "one-line" this function, very hard to read

achampagnedev replied:
this was just fixed I by the latest push :)

achampagnedev commented:
![](https://github.trello.services/images/mini-trello-icon.png) [GIG - Royalty+price input - Negative value](https://trello.com/c/eErTVlGD/391-gig-royaltyprice-input-negative-value)

achampagnedev commented:
@achampagnedev Missing the quantity input in `Offer/Form/Bid`

achampagnedev commented:
The summary of `Sales/Direct/Form` needs a similar fix for BigNumber crash as `Offer/Form/Summary`

achampagnedev committed:
Components NumberInput Validation

achampagnedev committed:
Templates NumberInput Validation

achampagnedev committed:
Inverted regex that checks for decimals for better validation UX

achampagnedev committed:
Fixed UI for min/max price filters

achampagnedev committed:
Updated changelog

achampagnedev commented:
@NicolasMahe This is ready to test. I think I covered most cases. I can see a potential issue with `currency.decimals` whenever present. It seems that either the number of decimals is not updated when switching currency but it's always `18`. Will take a look in the morning. Marketplace Template PR is also available and can be compared with this one for anything that can't be tested well within the storybook scope.

NicolasMahe commented on packages/components/src/Token/Form/Create.tsx:
```diff
@@ -298,18 +304,27 @@ const TokenFormCreate: FC<Props> = ({
-            step={0.01}
             allowMouseWheel
             w="full"
+            step={0.5}
```
Let's keep the existing step:
```suggestion
            step={0.01}
```

achampagnedev replied:
Updated

NicolasMahe commented on packages/templates/src/home/Explorer.tsx:
```diff
@@ -504,53 +505,109 @@ export const Template: VFC<
+                          placeholder={t('explore.form.min-price.placeholder')}
+                          {...register('minPrice', {
+                            validate: (value) => {
+                              if (!value) return
+                              const splitValue = value.toString().split('.')
```
Why only the validations in this file need to convert the value to a string?

achampagnedev replied:
Seems if these are not numbers it creates issues with the filters logic. I tried to change it to strings but seems to create more problems, might be better to create a different issue if needed

NicolasMahe commented on packages/components/src/Offer/Form/Checkout.tsx:
```diff
@@ -151,6 +148,20 @@ const OfferFormCheckout: FC<Props> = ({
+                        parseInt(offer.availableQuantity, 10)
+                    ) {
+                      return t('offer.form.checkout.validation.in-range', {
+                        available: parseInt(offer.availableQuantity, 10),
```
Seems the translation wants `quantityAvailable`:
```suggestion
                        quantityAvailable: parseInt(offer.availableQuantity, 10),
```

achampagnedev replied:
normalized all translations

NicolasMahe commented on packages/components/src/Offer/Summary.tsx:
```diff
@@ -19,7 +19,12 @@ const Summary: FC<
+    try {
+      return price.mul(quantity)
+    } catch {
+      console.error(`Cannot parse quantity ${price} as BigNumber`)
```
```suggestion
      console.error(`Cannot parse quantity ${quantity} as BigNumber`)
```

NicolasMahe commented on packages/components/src/Sales/Direct/Form.tsx:
```diff
@@ -124,15 +124,45 @@ const SalesDirectForm: VFC<Props> = ({
     return priceUnit.mul(feesPerTenThousand).div(10000)
   }, [price, priceUnit, feesPerTenThousand])
 
+  const amountFeesWithQuantity = useMemo(() => {
```
Why not parsing the quantity to a BigNumber in a single memo and use the value of this memo to do the calculation at the bottom of this file? That will require a lot less duplication
```tsx
  const quantityBN = useMemo(() => {
    if (!quantity) return BigNumber.from(0)
    try {
      return BigNumber.from(quantity)
    } catch {
      console.error(`Cannot parse quantity as BigNumber with value ${quantity}`)
      return BigNumber.from(0)
    }
  }, [quantity])
```

achampagnedev replied:
Updated

NicolasMahe commented on packages/components/locales/zh-cn/components.json:
```diff
@@ -374,7 +377,9 @@
-          "required": "此字段为必填字段"
+          "required": "此字段为必填字段",
+          "integer": "此字段必须是整数",
+          "in-range": "此字段必须介于 1 和 {{ availableQuantity }} 之间"
```
```suggestion
          "in-range": "此字段必须介于 1 和 {{ quantityAvailable }} 之间"
```

achampagnedev replied:
normalized all translations

NicolasMahe commented on packages/components/src/Offer/Form/Bid.tsx:
```diff
@@ -310,6 +312,19 @@ const OfferFormBid: FC<Props> = (props) => {
+                      parseInt(value, 10) > parseInt(props.supply, 10)
+                    ) {
+                      return t('offer.form.bid.validation.in-range', {
+                        quantityAvailable: parseInt(props.supply, 10),
```
This variable or the translation is wrong also. Can you settle on one name for this variable and use it everywhere.
Maybe something like `max`?
Same comment of `maxRoyalties` in translations, just use `max` all the time. The simpler the better.

achampagnedev replied:
normalized all translations

NicolasMahe commented on packages/components/src/Offer/Form/Bid.tsx:
```diff
@@ -243,9 +243,7 @@ const OfferFormBid: FC<Props> = (props) => {
         <InputGroup>
           <NumberInput
             clampValueOnBlur={false}
-            min={0}
```
I think we should keep the `min={0}` property to not let the stepper goes negative.

achampagnedev replied:
Added min and max where needed to prevent stepper to go below/above

NicolasMahe commented on packages/components/src/Offer/Form/Bid.tsx:
```diff
@@ -243,9 +243,7 @@ const OfferFormBid: FC<Props> = (props) => {
-            precision={currency.decimals}
             allowMouseWheel
             w="full"
             onChange={(x) => setValue('bid', x)}
```
This comment is about the next line,
why is this property format doing?
```
format={(e) => e.toString()}
```

achampagnedev replied:
Removed as it doesn't seem to affect functionality

achampagnedev committed:
PR change requests update

achampagnedev committed:
Reverted change on explorer types for number input and filters

achampagnedev commented:
@NicolasMahe Updated with requested updates.

- Normalized translation variables
- Added back min/max to prevent stepper from going below or above value of validation (except in the case of prices where min is 0 but will still require the user to enter a value above 0)
- Removed duplication with the quantity BigNumber and using a single memo
- Removed the format as it seems like the input already returns a `StringOrNumber` but the result seems to always be a string. the `onChange` returns a string so the `setValue` is still set as a string
- The validation on explore page needing to be converted `toString()` seems to affect the filters if not passed as a number in the files above filter logic. I tried to change the logic around to string but had issues with the filters not working. I suppose we can create an other issue for it if we want to fix this as well but I think filters might change soon with collections.

Let me know if I missed anything or didn't solve the right way

NicolasMahe commented on packages/templates/src/home/Explorer.tsx:
```diff
@@ -504,53 +505,109 @@ export const Template: VFC<
+                          placeholder={t('explore.form.min-price.placeholder')}
+                          {...register('minPrice', {
+                            validate: (value) => {
+                              if (!value) return
```
I think this should be removed because it bypass the other validations if the value is empty or zero.
Only this file have this logic, validation in other files do not have this.
Based on the typescript type, it should be:
```
if (value === null) return
```

NicolasMahe commented on packages/components/src/Offer/Form/Bid.tsx:
```diff
@@ -297,19 +303,30 @@ const OfferFormBid: FC<Props> = (props) => {
           <InputGroup>
             <NumberInput
```
If I put a point, it trigger a BigNumber error in the useMemo of totalPrice.
Need to check all other inputs for similar issue as well.

NicolasMahe commented:
Let's make this modif on https://github.com/liteflow-labs/starter-kit/pull/89 first so we can test them easily and then bring them back to the lib.
I feel a lot of issues are still present because on the lack of testing because the current lib organisation.

achampagnedev committed:
Fixed changes that shouldn't cause a diff

NicolasMahe commented:
@achampagnedev Now that PR https://github.com/liteflow-labs/starter-kit/pull/89 is merged, you can bring the same modif to this PR

achampagnedev committed:
Updated with latest changes

achampagnedev committed:
Fixed parseBigNumber

achampagnedev committed:
Fix storybook

NicolasMahe commented on packages/hooks/CHANGELOG.md:
```diff
@@ -77,7 +77,7 @@
   - `useFormatAddress`: kept helper `formatAddress` [#2107](https://github.com/liteflow-labs/nft/pull/2107)
   - `useErrorFormatter`: kept helper `formatError` [#2107](https://github.com/liteflow-labs/nft/pull/2107)
-  - `useParsePrice`: kept helper `parsePrice` [#2107](https://github.com/liteflow-labs/nft/pull/2107)
+  - `useParsePrice`: kept helper `parseBigNumber` [#2107](https://github.com/liteflow-labs/nft/pull/2107)
```
You went a bit too far with the renaming! Don't change past changelog please

NicolasMahe commented on packages/hooks/src/index.ts:
```diff
@@ -90,7 +90,7 @@ export {
   formatSocial,
   isSameAddress,
   removeEmptyFromObject,
-  parsePrice,
```
please keep `parsePrice` to not add any breaking change even if it not used internally

NicolasMahe reviewed (commented):
No diff found compared to https://github.com/liteflow-labs/marketplace-template/pull/89/ except 2 small things. Almost there!

NicolasMahe committed:
revert removing parsePrice

NicolasMahe committed:
revert modif on changelog
